File: src/chart_xkcd/bar.py
"""Bar chart."""

from typing import Any
from .charts import _AxisChart, _check_labels_datasets


class Bar(_AxisChart):
    """Bar chart with categorical x-axis.

    By default renders only ``datasets[0]``. Set ``grouped=True`` in options
    to render all datasets as side-by-side bars within each category.

    Args:
        title: Chart title.
        x_label: Label for the x-axis.
        y_label: Label for the y-axis.
        labels: Category names for the x-axis.
        datasets: List of dataset dicts, each with a ``data`` key
            (list of numeric values) and an optional ``label`` key.
            Only the first dataset is used unless ``grouped=True``.
        options: Dict of chart options.

    Options:

    - ``backgroundColor`` (str): Background color (default ``'white'``).
    - ``dataColors`` (list[str]): Bar colors.
    - ``fontFamily`` (str): Font family (default ``'xkcd'``).
    - ``grouped`` (bool): Render all datasets as grouped side-by-side bars
      (default False).
    - ``strokeColor`` (str): Axis/border color (default ``'black'``).
    - ``unxkcdify`` (bool): Disable hand-drawn style (default False).
    - ``yMin`` (float): Minimum y-axis value (default 0).
    - ``yMax`` (float): Maximum y-axis value (default data max).
    - ``yTickCount`` (int): Number of y-axis ticks (default 3).

    Example — simple bar::

        Bar(
            title="Monthly Sales",
            x_label="Month",
            y_label="Revenue",
            labels=["Jan", "Feb", "Mar"],
            datasets=[{"data": [10, 20, 30]}],
        )

    Example — grouped bar::

        Bar(
            title="Quarterly Sales",
            x_label="Quarter",
            y_label="Units",
            labels=["Q1", "Q2", "Q3"],
            datasets=[
                {"data": [10, 20, 30], "label": "Widgets"},
                {"data": [5, 15, 10], "label": "Gadgets"},
            ],
            options={"grouped": True},
        )
    """

    def __init__(
        self,
        *,
        title: str | None = None,
        x_label: str | None = None,
        y_label: str | None = None,
        labels: Any,
        datasets: Any,
        options: Any = None,
    ):
        _check_labels_datasets(labels, datasets, "Bar")
        data = {"labels": list(labels), "datasets": list(datasets)}
        super().__init__(
            title=title, x_label=x_label, y_label=y_label, data=data, options=options
        )

File: js/src/Bar.js
import { select, mouse, event as d3Event } from 'd3-selection';
import { scaleBand, scaleLinear } from 'd3-scale';

import addAxis from './utils/addAxis';
import { tooltipPositionType } from './components/Tooltip';
import config from './config';
import {
  applyDefaults, setupMargin, resolveFilterAndFont,
  createSvgEl, setupChartGroup, createTooltip,
} from './utils/initChart';

/**
 * Simple bar chart with one dataset.
 *
 * Renders vertical bars with an xkcd hand-drawn style filter.
 * Supports hover tooltips, click selection, and shift-click
 * multi-selection.
 *
 * @param {SVGElement} svg - Target SVG element (usually inside a sized container).
 * @param {Object} params
 * @param {string} [params.title] - Chart title displayed above the plot.
 * @param {string} [params.xLabel] - Label for the x-axis.
 * @param {string} [params.yLabel] - Label for the y-axis.
 * @param {Object} params.data
 * @param {string[]} params.data.labels - Category labels for the x-axis.
 * @param {Object[]} params.data.datasets - Array with a single dataset object
 *   containing a `data` array of numeric values and an optional `label`.
 * @param {Object} [params.options] - See `applyDefaults` for common options.
 */
class Bar {
  constructor(svg, {
    title, xLabel, yLabel, data: { labels, datasets }, options,
  }) {
    this.options = applyDefaults({
      yTickCount: config.defaultTickCount,
      ...options,
    });
    this.title = title;
    this.xLabel = xLabel;
    this.yLabel = yLabel;
    this.data = { labels, datasets };

    const margin = setupMargin({ title, xLabel, yLabel });
    const { filter, fontFamily } = resolveFilterAndFont(this.options, false);
    this.filter = filter;
    this.fontFamily = fontFamily;
    this.svgEl = createSvgEl(svg, { fontFamily, backgroundColor: this.options.backgroundColor });
    const { chart, width, height } = setupChartGroup(
      this.svgEl, margin, { title, xLabel, yLabel, strokeColor: this.options.strokeColor },
    );
    this.chart = chart;
    this.width = width;
    this.height = height;
    this.margin = margin;
    this.render();
  }

  render() {
    const tooltip = createTooltip(this.svgEl, this.options);

    const xScale = scaleBand()
      .range([0, this.width])
      .domain(this.data.labels)
      .padding(config.bandPadding);

    const allData = this.data.datasets
      .reduce((pre, cur) => pre.concat(cur.data), []);

    const yScale = scaleLinear()
      .domain([0, Math.max(...allData)])
      .range([this.height, 0]);

    const graphPart = this.chart.append('g');

    addAxis.xAxis(graphPart, {
      xScale,
      tickCount: config.defaultTickCount,
      moveDown: this.height,
      fontFamily: this.fontFamily,
      unxkcdify: this.options.unxkcdify,
      stroke: this.options.strokeColor,
    });
    addAxis.yAxis(graphPart, {
      yScale,
      tickCount: this.options.yTickCount,
      fontFamily: this.fontFamily,
      unxkcdify: this.options.unxkcdify,
      stroke: this.options.strokeColor,
    });

    graphPart.selectAll('.xkcd-chart-bar')
      .data(this.data.datasets[0].data)
      .enter()
      .append('rect')
      .attr('class', 'xkcd-chart-bar')
      .attr('x', (d, i) => xScale(this.data.labels[i]))
      .attr('width', xScale.bandwidth())
      .attr('y', (d) => yScale(d))
      .attr('height', (d) => this.height - yScale(d))
      .attr('fill', 'none')
      .attr('pointer-events', 'all')
      .attr('stroke', this.options.strokeColor)
      .attr('stroke-width', config.barStrokeWidth)
      .attr('rx', config.barCornerRadius)
      .attr('filter', this.filter)
      .on('mouseover', (d, i, nodes) => {
        select(nodes[i]).attr('fill', this.options.dataColors[i]);
        tooltip.show();
      })
      .on('mouseout', (d, i, nodes) => {
        select(nodes[i]).attr('fill', 'none');
        tooltip.hide();
      })
      .on('click', (d, i) => {
        if (this.options.onSelect) {
          this.options.onSelect({ index: i, label: this.data.labels[i], value: d }, d3Event.shiftKey);
        }
      })
      .on('mousemove', (d, i, nodes) => {
        const tipX = mouse(nodes[i])[0] + this.margin.left + config.tooltipMouseOffset;
        const tipY = mouse(nodes[i])[1] + this.margin.top + config.tooltipMouseOffset;
        tooltip.update({
          title: this.data.labels[i],
          items: [{
            color: this.options.dataColors[i],
            text: `${this.data.datasets[0].label || ''}: ${d}`,
          }],
          position: {
            x: tipX,
            y: tipY,
            type: tooltipPositionType(tipX, tipY, this.width, this.height),
          },
        });
      });
  }
}

export default Bar;
